#! /bin/sh /usr/share/dpatch/dpatch-run
## Full Linux crash by mouse movement + cursor change with fglrx drivers for ATI graphics.dpatch by  <root@debian.example.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Work around buggy ATI drivers by Henri Beauchamp

@DPATCH@

diff -urN sl117/indra/llwindow/llwindowsdl.cpp sl117-patched/indra/llwindow/llwindowsdl.cpp
--- sl117/indra/llwindow/llwindowsdl.cpp	2007-06-13 19:10:26.000000000 +0200
+++ sl117-patched/indra/llwindow/llwindowsdl.cpp	2007-06-21 17:55:05.000000000 +0200
@@ -57,6 +57,9 @@
 
 const S32	MAX_NUM_RESOLUTIONS = 32;
 
+// static variable for ATI mouse cursor crash-bug work around:
+static int ATIbug = FALSE;
+
 //
 // LLWindowSDL
 //
@@ -2247,6 +2250,10 @@
 
 void LLWindowSDL::setCursor(ECursorType cursor)
 {
+	if (ATIbug) {
+		return;
+	}
+
 	if (mCurrentCursor != cursor)
 	{
 		if (cursor < UI_CURSOR_COUNT)
@@ -2316,6 +2323,10 @@
 	mSDLCursors[UI_CURSOR_TOOLPAY] = makeSDLCursorFromBMP("toolpay.BMP",0,0);
 	mSDLCursors[UI_CURSOR_TOOLOPEN] = makeSDLCursorFromBMP("toolopen.BMP",0,0);
 	mSDLCursors[UI_CURSOR_PIPETTE] = makeSDLCursorFromBMP("lltoolpipette.BMP",2,28);
+
+	if (getenv("LL_ATI_MOUSE_CURSOR_BUG") != NULL) {
+		ATIbug = TRUE;
+	}
 }
 
 void LLWindowSDL::quitCursors()
diff -urN sl117/indra/newview/linux_tools/wrapper.sh sl117-patched/indra/newview/linux_tools/wrapper.sh
--- sl117/indra/newview/linux_tools/wrapper.sh	2007-06-13 19:10:29.000000000 +0200
+++ sl117-patched/indra/newview/linux_tools/wrapper.sh	2007-06-21 17:58:58.000000000 +0200
@@ -43,6 +43,11 @@
 ## - Avoids an often-buggy X feature that doesn't really benefit us anyway.
 export SDL_VIDEO_X11_DGAMOUSE=0
 
+## - Work around the ATI mouse cursor crash bug:
+if lsmod | grep fglrx &>/dev/null ; then
+	export LL_ATI_MOUSE_CURSOR_BUG=x
+fi
+
 ## Nothing worth editing below this line.
 ##-------------------------------------------------------------------
 
